### PR TITLE
feat: pipeline visualizer

### PR DESCRIPTION
**Pull request by adong**
## Context

Issue: https://github.com/screwdriver-cd/screwdriver/issues/1056

## Objective

Add a new route under tools to allow users to search and navigate pipelines. So all relevant connected pipelines are shown in a single workflow graph.

Note: currently the depth of the graph is 1, meaning only the current _upstream_ and _downstream_ of a given pipeline are shown.

See screenshot:
image

image

...TESTS PENDING...

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by tkyi** [comment]
Is it possible to show the `repo name -> pipeline number` relationship somehow? Could be hard for people to figure out what they're looking at otherwise

**Comment by adong** [comment]
> Is it possible to show the `repo name -> pipeline number` relationship somehow? Could be hard for people to figure out what they're looking at otherwise

Yeah, good idea. I could append the pipeline number to the repo name.

**Comment by tkyi** [comment]
@adong I'm not sure if it's the optimal way to show the information, but it's good enough for now. We can improve on that later

File: app/components/pipeline-viz-chart/component.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    externalTrigger: {
+      pipelineId: 0
+    }
+  },
+
+  workflowGraph: computed('pipeline', function workflowGraph() {
+    return this.pipeline.workflowGraph;
+  }),
+  startFrom: computed('pipeline', function startFrom() {
+    // return this.pipeline.startFrom;
```
Comment (tkyi) [inline app/components/pipeline-viz-chart/component.js:36]: commented code

File: app/pipeline-visualizer/controller.js [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+import EmberObject, { getWithDefault } from '@ember/object';
+import { all, Promise as EmberPromise } from 'rsvp';
+import { copy } from 'ember-copy';
+import $ from 'jquery';
+import { debounce, later } from '@ember/runloop';
+
+import { removeBranch } from 'screwdriver-ui/utils/graph-tools';
+
+/**
+ * Check weather if given job name is an external trigger
```
Comment (tkyi) [inline app/pipeline-visualizer/controller.js:12]:
```suggestion
 * Check whether given job name is an external trigger
```

File: public/assets/svg/pipeline-vis-display.svg [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                        </g>
+                    </g>
+                    <!-- <rect id="Rectangle" fill="url(#linearGradient-1)" x="0.249266862" y="0" width="25.9237537" height="181.466276"></rect> -->
+                    <!-- <rect id="Rectangle-Copy" fill="url(#linearGradient-1)" transform="translate(326.290323, 90.733138) scale(-1, 1) translate(-326.290323, -90.733138) " x="313.328446" y="0" width="25.9237537" height="181.466276"></rect> -->
```
Comment (tkyi) [inline public/assets/svg/pipeline-vis-display.svg:56]: commented code

**Comment by pritamstyz4ever** [comment]
> > Is it possible to show the `repo name -> pipeline number` relationship somehow? Could be hard for people to figure out what they're looking at otherwise
> 
> Yeah, good idea. I could append the pipeline number to the repo name.
> 
> How about this?
> image

Is that a dropdown to search for pipelines, can we not search by pipelineId?

**Comment by adong** [comment]
> > > Is it possible to show the `repo name -> pipeline number` relationship somehow? Could be hard for people to figure out what they're looking at otherwise
> > 
> > 
> > Yeah, good idea. I could append the pipeline number to the repo name.
> > How about this?
> > image
> 
> Is that a dropdown to search for pipelines, can we not search by pipelineId?

No. Currently, our API does not support search by id, only text. Users can use url like  `https://cd.screwdriver.cd/pipeline-visualizer?pipelineId=73` if they know pipeline id ahead of time.

**Review by klu909 — APPROVED** [review]

**Pull request merged**